chapters/part3/hyperparameter-optimization-with-optuna.md

```diff
@@ -0,0 +1,426 @@
+
+```python
+import optuna
+
+
+def objective(trial: optuna.Trial) -> float:
+    trial.suggest_int("embedding_dim", 32, 256)
+    trial.suggest_int("max_filter_size", 2, 6)
+    trial.suggest_int("num_filters", 32, 256)
+    trial.suggest_int("output_dim", 32, 256)
+    trial.suggest_float("dropout", 0.0, 0.8)
+    trial.suggest_float("lr", 5e-3, 5e-1, log=True)
+
+    executor = optuna.integration.allennlp.AllenNLPExecutor(
+        trial=trial,  # trial object
+        config_file="./config/imdb_optuna.jsonnet",  # path to jsonnet
+        serialization_dir=f"./result/optuna/{trial.number}",
+        metrics="best_validation_accuracy"
+    )
+    return executor.run()
+```
+
+After defining search spaces using `trial.suggest_int` and `trial.suggest_float`, the `trial` object should be passed to `AllenNLPExecutor`.
+The `trial` object holds all suggested values after defining search spaces using `trial.suggest_int` and `trial.suggest_float`,  and it should be pass to `AllenNLPExecutor`.
+`AllenNLPExecutor` takes four required arguments; `trial` (Optuna's object), `config_file` (path to a model configuration),
+`serialization_dir` (directory for saving model snapshot, log, etc.), and `metrics` you want to optimize.
+In the above example, we create an instance of `AllenNLPExecutor` as `executor`.
+Once the `executor` instance is created, training is started with `executor.run()`.
+In each trial step in optimization, the objective function is called and does the following steps:
+
+1. Train a model (`executor.run()`)
+2. Return a target metric on validation data (`executor.run()` returns the specified metric)
+
+Now, we finished defining objective function.
+Let's write Optuna stuff for launching optimization!
+In Optuna, we create a study object and pass the objective function to the `optimize()` method as follows.
+You can specify something; a way to save a result of optimization, a sampler for searching hyperparameters (`TPESampler` is based on Bayesian Optimization),
+direction for optimizing (maximize or minimize), number of jobs for distributed training, or timeout.
+
+```python
+if __name__ == '__main__':
+    study = optuna.create_study(
+        storage="sqlite:///result/trial.db",  # save results in DB
+        sampler=optuna.samplers.TPESampler(seed=24),
+        study_name="optuna_allennlp",
+        direction="maximize",
+    )
+
+    timeout = 60 * 60 * 10  # timeout (sec): 60*60*10 sec => 10 hours
+    study.optimize(
+        objective,
+        n_jobs=1,  # number of processes in parallel execution
+        n_trials=30,  # number of trials to train a model
+        timeout=timeout,  # threshold for executing time (sec)
+    )
+```
+
+</exercise>
+
+<exercise id="3" title="Results of Hyperparameter Optimization">
+
+If `study.optimize` successfully runs, `trial.db` would be created in the directory `result`.
+
+[Tip] You can change a names of database file and directory by changing the value of `storage` in `optuna.create_study`.
```

Comment:
This tip may not be necessary if this text immediately follows the example launch script above, where we already mentioned that we can vary the storage location.

Reply:
Thank you, I removed the tip since [your suggestion](https://github.com/allenai/allennlp-guide/pull/138#discussion_r472491586) explained this well. b690276